Add metadata endpoint

BarryPiccinni opened the pull request.

BarryPiccinni committed:
Add metadata endpoint

BarryPiccinni committed:
Added metadata to openapi-src

BarryPiccinni committed:
Add get answer endpoint

BarryPiccinni commented on openapi/src/json-schemas/api/resources/metadata-resource-v1.0.0.json:
```diff
@@ -0,0 +1,35 @@
+                    "const": "metadata"
+                },
+                "id": {
+                    "const": "0"
```
I have changed this to be a questionnaire Id as suggested.

armoj commented on openapi/openapi.json:
```diff
@@ -3362,6 +3362,229 @@
+                                                                    },
+                                                                    {
+                                                                        "properties": {
+                                                                            "type": {
```
is it worth employing `"additionalProperties": true`, or `patternProperties` here or anything. Make it is strict/loose as needed and still allow arbitrary key/values

armoj reviewed (changes requested):
What is the logic behind the version 1.0.0/2.0.0 of the resource for meta? I remember it being used for the dataset endpoint due to there needing to be 2 version of the response for that endpoint, but not sure what its purpose is here. This is just a question, not a request to alter it

sinclairs commented on openapi/src/openapi-src.json:
```diff
@@ -228,6 +228,101 @@
+                    {
+                        "$ref": "#/components/parameters/sectionId"
+                    },
+                    {
```
Given the all answers endpoint isn't versioned, it might be worth leaving off the versioning for the first pass

sinclairs commented:
> What is the logic behind the version 1.0.0/2.0.0 of the resource for meta? I remember it being used for the dataset endpoint due to there needing to be 2 version of the response for that endpoint, but not sure what its purpose is here. This is just a question, not a request to alter it

https://cloud.google.com/blog/products/api-management/api-design-which-version-of-versioning-is-right-for-you

It allows us to change the resource over time. Although I think the guidance above does suggest that it might not always be useful.

BarryPiccinni committed:
Update openApi responses with all 4xx responses

armoj commented on openapi/src/json-schemas/api/resources/metadata-resource-v1.0.0.json:
```diff
@@ -0,0 +1,35 @@
+                "id": {
+                    "const": "0"
+                },
+                "attributes": {
```
In order to avoid having to use quotes for object property names?

BarryPiccinni committed:
Update examples for meta endpoint

BarryPiccinni committed:
Correct examples and filter content

BarryPiccinni commented:
@sinclairs I've fixed the examples in my latest push. I've also added a second filter component as it was specific to progress entries and so didn't make sense when including it in the meta endpoint

BarryPiccinni committed:
fix typo

BarryPiccinni committed:
Run openAPI build

BarryPiccinni committed:
Update timezone validation to validate only ISO8601 dates

BarryPiccinni committed:
run OpenAPI build

BarryPiccinni committed:
Fix typo

BarryPiccinni committed:
Add DAL for new endpoints

BarryPiccinni committed:
Update user answer keys

sinclairs commented on openapi/src/json-schemas/models/definitions/timestamp.json:
```diff
@@ -0,0 +1,6 @@
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Timestamp",
+    "type": "string",
+    "pattern": "^(?:[1-9]\\d{3}-(?:(?:0[1-9]|1[0-2])-(?:0[1-9]|1\\d|2[0-8])|(?:0[13-9]|1[0-2])-(?:29|30)|(?:0[13578]|1[02])-31)|(?:[1-9]\\d(?:0[48]|[2468][048]|[13579][26])|(?:[2468][048]|[13579][26])00)-02-29)T(?:[01]\\d|2[0-3]):[0-5]\\d:[0-5]\\d(?:\\.\\d{1,9})?(?:Z|[+-][01]\\d:[0-5]\\d)$"
```
JSON Schema already has this baked in e.g. https://github.com/CriminalInjuriesCompensationAuthority/q-templates-application/blob/b40fa7ce73d52c3051acd1b1f1db8fffafa30a61/src/lib/resource/sections/applicant-enter-your-date-of-birth.js#L59

Do we need this regex?

BarryPiccinni replied:
I was quite pleased with how complex it was, but no we don't need it! Have changed it to the JSON schema format.

BarryPiccinni committed:
Swap datetime regex for json schema format

armoj commented:
Adding here for visibility:

For the metadata DCS implementation, in the interest of separation of concerns... I am thinking it would be beneficial to separate out the metadata functionality in to its own directory (similarly to the dataset stuff).

`/questionnaire/metadata`
`/questionnaire/metadata/route.js` Have your endpoints defined in here and link to it [like this](https://github.com/CriminalInjuriesCompensationAuthority/data-capture-service/blob/ddae84a802ec8a85543c8b2f2c4cc2c59497de12/questionnaire/routes.js#L41) in the questionnaire routes.js
`/questionnaire/metadata/metadata-service.js` (have the logic to get/build the resource(s) in here. You can also then have a util/helper function that will add 31 days to the created date with no need to "one line" it. Better for testability also.

BarryPiccinni committed:
Extract metadata into own folder

BarryPiccinni committed:
Remove metadata tests from questionnaire service test file

BarryPiccinni commented:
> You can also then have a util/helper function that will add 31 days to the created date with no need to "one line" it. Better for testability also.

I didn't do this as the `expires` function is only a first pass until we store it in the DB and reference it as we do the rest of the metadata.

armoj commented on questionnaire/questionnaire-dal.js:
```diff
@@ -177,6 +177,34 @@ function questionnaireDAL(spec) {
+            throw err;
+        }
+
+        return result;
```
.map() this to remove all other redundant data in the response. Just return an array of objects

BarryPiccinni committed:
Amend db return value

armoj commented on questionnaire/routes.js:
```diff
@@ -105,6 +107,18 @@ router
+                req.params.questionnaireId,
+                req.params.sectionId
+            );
+            res.status(200).json(response);
```
change to:
````js
res.status(200).json({
    data: response
});
````

BarryPiccinni committed:
Correct return value of get-section-by-id function

sinclairs commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
     "name": "data-capture-service",
-    "version": "6.2.6",
```
As this is adding new functionality in a backwards compatible way, this should be a minor release rather than a patch.

If it was behind a toggle the patch would apply

BarryPiccinni replied:
This is being merged into an RC, so I've removed all versioning for now. We can version the RC at deployment.

armoj commented on questionnaire/questionnaire-dal.js:
```diff
@@ -177,6 +177,34 @@ function questionnaireDAL(spec) {
+                }
+            } else {
+                result = await db.query(
+                    "SELECT id, questionnaire -> 'meta' -> 'questionnaireDocumentVersion' AS \"questionnaire-version\", created, modified, submission_status, questionnaire -> 'answers' -> 'user' -> 'user-id' AS \"user-id\" FROM questionnaire"
```
should this else block use the same pattern as the preceding if block above it?: get result, throw if rowCount is 0. Then return at the bottom of the function if it falls through to it

BarryPiccinni replied:
Well the thinking was to get to this `else` branch the user has asked for all metadata. The result is either some metadata or no metadata. They've not provided any parameters that would generate an error.

sinclairs commented on questionnaire/routes.test.js:
```diff
@@ -188,3 +188,44 @@ describe('Issue: https://github.com/cdimascio/express-openapi-validator/issues/7
+describe('/questionnaires/meta?filter[userId]=foobar', () => {
+    describe('get', () => {
+        describe('200', () => {
+            it('should replace JSON Pointers with the value they reference', async () => {
+                // eslint-disable-next-line global-require
+                const mockResponse = require('./test-fixtures/res/questionnaireCompleteWithCRN.json');
+
+                // mock the DAL db integration
+                jest.doMock('./questionnaire-dal.js', () =>
+                    // return a modified factory function, that returns an object with a method, that returns a valid created response
+                    jest.fn(() => ({
+                        getQuestionnaire: () => mockResponse,
+                        getQuestionnaireModifiedDate: () => undefined
+                    }))
+                );
+
+                const tokens = {
+                    'read:progress-entries':
+                        'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJkYXRhLWNhcHR1cmUtc2VydmljZSIsImlzcyI6IiQuYXVkIiwianRpIjoiNDUwMzE2ZTYtNDFhNS00MGRjLWI3NTUtMzA2ZGQ2M2FlMDhiIiwic3ViIjoiY2ljYS13ZWIiLCJzY29wZSI6InJlYWQ6cHJvZ3Jlc3MtZW50cmllcyIsImlhdCI6MTU2NTc5NzE1MH0.fF6Ln7GZmq-R36N-Avuo_a_8Jj5-wla17x0552XnMbE'
+                };
+                // app has an indirect dependency on questionnaire-dal.js, require it after
+                // the mock so that it references the mocked version
+                // eslint-disable-next-line global-require
+                const app = require('../app');
+                const res = await request(app)
+                    .get(
+                        '/api/v1/questionnaires/285cb104-0c15-4a9c-9840-cb1007f098fb/progress-entries?filter[position]=current'
+                    )
+                    .set('Authorization', `Bearer ${tokens['read:progress-entries']}`);
+                const confirmation =
+                    res.body.included[0].attributes.properties.confirmation.description;
+
+                expect(confirmation).toMatch(/19\\751194/);
+                expect(confirmation).toMatch(/Barry\.foo@foo\.com/);
+                expect(res.statusCode).toBe(200);
+                expect(res.type).toBe('application/vnd.api+json');
+            });
+        });
+    });
+});
```
Can this test live with the other metadata routing tests? Also, I'm not quite following what the metadata bit of this test is?

sinclairs commented on questionnaire/metadata/metadata-service.test.js:
```diff
@@ -0,0 +1,49 @@
+                createQuestionnaireDAL: () => ({
+                    getQuestionnaireMetadata: () => {
+                        return [
+                            {
+                                id: '6fbc0958-e70c-4e37-ab65-12b96e8d37c1',
+                                'questionnaire-version': '4.2.0',
+                                created: '2023-01-01T10:42:24.016Z',
+                                modified: '2023-01-01T10:42:49.604Z',
+                                submission_status: 'NOT_STARTED',
+                                'user-id': 'abc123'
+                            }
+                        ];
+                    }
+                })
+            });
+
+            const actual = await metadataService.getMetadata();
+
+            const expected = {
+                data: [
+                    {
+                        type: 'metadata',
+                        id: '6fbc0958-e70c-4e37-ab65-12b96e8d37c1',
+                        attributes: {
+                            'questionnaire-id': '6fbc0958-e70c-4e37-ab65-12b96e8d37c1',
+                            'questionnaire-document-version': '4.2.0',
+                            created: '2023-01-01T10:42:24.016Z',
+                            modified: '2023-01-01T10:42:49.604Z',
+                            state: 'NOT_STARTED',
+                            'user-id': 'abc123',
+                            expires: '2023-02-01T00:00:00.000Z'
```
The keys in the response seem to differ from those returned from the DAL e.g. `submission_status` to `state`. I thought they'd be the same? `submission-status` is more descriptive

sinclairs commented on questionnaire/metadata/metadata-routes.js:
```diff
@@ -0,0 +1,20 @@
+
+const router = express.Router();
+
+router.route('/meta').get(permissions('read:questionnaires'), async (req, res, next) => {
```
For consistency, should all "meta" refs be replaced with "metadata"

sinclairs reviewed (changes requested):
Looking good, just a few small things

BarryPiccinni commented on questionnaire/questionnaire-dal.js:
```diff
@@ -177,6 +177,34 @@ function questionnaireDAL(spec) {
+                        {
+                            name: 'ResourceNotFound'
+                        },
+                        `User "${query.filter.userId}" not found`
```
I feel like the error is derived from the parameter here. Metadata can exist or there is no metadata - both are valid results - but a dodgy userId should throw an error. We don't currently distinguish between a dodge userId and 0 rows tho, so I can see both sides of this.

BarryPiccinni committed:
Add changes requested at PR

BarryPiccinni committed:
Fix failing tests

BarryPiccinni committed:
Fix questionnaire-document-version return value
